### PR TITLE
feat: adhoc network support

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #598 

### How I did it

* Handle when user puts URL at end of ecosystem / network choice string, e.g.:
```
--network ethereum:mainnet:https://foo.bar
--network ::https://foo.bar
```
* Handle when user omits ecosystem and network entirely and use "adhoc"-like settings (geth provider reliant)
```
--network https://foo.bar
```

Here are examples:

Using ecosystem and network:

```
(angel39) ➜  ape-demo-project git:(main) ✗ ape console --network ethereum:mainnet:https://<omit>:<omit>@erigon<omit>/
INFO: Connecting to existing Erigon node at '<omit>'

In [1]: provider
Out[1]: <geth chain_id=1>

In [2]: provider.chain_id
Out[2]: 1

In [3]: provider.get_block("latest")
Out[3]: Block(num_transactions=135, hash=HexBytes('0xda1eb42477d29acdbd2e3bd2c6a517d66b687bc12a63efaca2bda8a39a3fa7ef'), number=15219795, parent_hash=HexBytes('0xf64ef0ffc7843277a29621b186673572b03bd3429cb9420744e427bcc9201979'), size=25401, timestamp=1658861362, gas_limit=29970705, gas_used=5695467, base_fee=16623710745, difficulty=11766754802216255, total_difficulty=54914317311543363044806)
```

And an example without using ecosystem or network:

```
(angel39) ➜  ape-demo-project git:(main) ✗ ape console --network https://rpc.ftm.tools
WARNING: Connecting Geth plugin to non-Geth client 'go-opera'.

In [1]: provider.chain_id
Out[1]: 250

In [2]: provider
Out[2]: <geth chain_id=250>

In [3]: provider.get_block("latest")
Out[3]: Block(num_transactions=6, hash=HexBytes('0x000218e9000011dcfa1c7fadf646fb78f357b9bdaee67ab14725721b616f498d'), number=43606522, parent_hash=HexBytes('0x000218e9000011cfa764771321513d1d92e8d3b7839b25b70169a836e24a0101'), size=14145, timestamp=1658861132, gas_limit=281474976710655, gas_used=13969043, base_fee=1000000000, difficulty=0, total_difficulty=0)
```

### How to verify it

See examples above!
Basically, try to connect to any random EVM-like network.
Anything that would work with a HTTPProvider in web3 should work this way as well.

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: docs/userguides/networks.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+```
+
+Additionally, if you want to connect to an unknown ecosystem or network, you can use the URI by itself.
+However, this is not recommended.
+
+```bash
+ape run script --network https://foo.bar
+```
+
+**The recommended approach is to find or build a plugin to have more native support.**
```
Comment (mikeshultz) [inline docs/userguides/networks.md:70]: suggestion: Might help to expand on this a little bit.  Above it sounds like a network param with a URI would just default to geth.  In what situation would building a plugin be beneficial to ad-hoc support?  Or do you mean an alternative to ad-hock connections?

File: src/ape/managers/networks.py [hunk 2]
```diff
@@ -119,6 +120,36 @@ def ecosystems(self) -> Dict[str, EcosystemAPI]:
...
+        """
+
+        geth_class = None
+        for plugin_name, (_, _, provider_class) in self.plugin_manager.providers:
+            if plugin_name == "geth":
+                geth_class = provider_class
+                break
+
+        if geth_class is None:
+            raise NetworkError("Core Geth plugin missing.")
```
Comment (mikeshultz) [inline src/ape/managers/networks.py:143]: question: Is this possible?  Unless you mod ape directly, is there a case where `ape_geth` could be unavailable?

File: src/ape/managers/networks.py [hunk 3]
```diff
@@ -303,11 +334,15 @@ def get_provider_from_choice(
                 provider_settings=provider_settings
             )
 
+        elif network_choice.startswith("http://") or network_choice.startswith("https://"):
```
Comment (mikeshultz) [inline src/ape/managers/networks.py:337]: thought: I wonder if other protocols should be supported, like `ipc://`.

File: tests/functional/test_cli.py [hunk 4]
```diff
@@ -49,3 +62,40 @@ def test_get_user_selected_account_unknown_type(runner, keyfile_account):
...
+    "network_input",
+    (
+        "something:else:https://127.0.0.1:4545",
+        "something:else:https://127.0.0.1",
+        "something:else:http://127.0.0.1:4545",
+        "something:else:http://127.0.0.1",
+        "something:else:http://foo.bar",
+        "something:else:https://foo.bar:8000",
+        "::https://foo.bar:8000",
+        ":else:https://foo.bar:8000",
```
Comment (mikeshultz) [inline tests/functional/test_cli.py:95]: suggestion(non-blocking): Could also include only URIs like `https://foo.bar:800`.

**Review by mikeshultz — COMMENTED** [review]
Appreciate you tackling this.  This is super useful to me.  Tried to add some useful commentary but overall LGTM!

File: docs/userguides/networks.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+When interacting with the blockchain, you will have to select a network.
+
+## Selecting a Network
+
+Commonly, you will use the `--network` option to configre your network during Ape commands.
+The following is a list of common Ape commands that can use the `--network` option:
+
+```bash
+ape test --network ethereum:local:foundry
+ape run deploy --network ethereum:mainnet:geth  # NOTE: Must be defined if using CLI script
```
Comment (NotPeopling2day) [inline docs/userguides/networks.md:12]: Maybe some more clarity on what must be defined

File: src/ape/managers/networks.py (hunk 2 above)
Comment (unparalleled-js) [reply src/ape/managers/networks.py:143]: It is possible when you have a messed up editable install of the core ape library.

File: src/ape/managers/networks.py (hunk 3 above)
Comment (unparalleled-js) [reply src/ape/managers/networks.py:337]: Yes, but that will require some additional effort to `ape-geth` so will save for another time most likely

File: tests/functional/test_cli.py (hunk 4 above)
Comment (unparalleled-js) [reply tests/functional/test_cli.py:95]: Done!

File: docs/userguides/networks.md (hunk 5 above)
Comment (unparalleled-js) [reply docs/userguides/networks.md]: Done! Had to add the scripts guide to explain more

File: src/ape/managers/networks.py (hunk 2 above)
Comment (mikeshultz) [reply src/ape/managers/networks.py:143]: Fair enough, defensive is good.  I'm unable to mark this as resolved, though.

File: src/ape/managers/networks.py (hunk 3 above)
Comment (mikeshultz) [reply src/ape/managers/networks.py:337]: ah, yeah, with web3.py you'd need to use `IPCProvider` and `ape_geth` only supports `HTTPProvider` currently.  :+1:

**Review by mikeshultz — DISMISSED** [review]
Thanks again for this

File: src/ape/api/networks.py [hunk 6]
```diff
@@ -497,6 +497,30 @@ class NetworkAPI(BaseInterfaceModel):
...
+        ethereum_class = None
+        for plugin_name, ecosystem_class in cls.plugin_manager.ecosystems:
+            if plugin_name == "ethereum":
+                ethereum_class = ecosystem_class
+                break
+
+        if ethereum_class is None:
+            raise NetworkError("Core Ethereum plugin missing.")
+
+        data_folder = cls.config_manager.DATA_FOLDER / "adhoc"
```
Comment (fubuloubu) [inline src/ape/api/networks.py:511]: Wonder if it might be best to write out to a temp folder for adhoc...

**Review by fubuloubu — DISMISSED** [review]
Really nice docs updates, feature looks really great too!

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**